### ISSUE TITLE
[BUG] GlassBreak() doesn't jump to other spawn states, only first state

### ISSUE THREAD

**Issue by deathzero021** [ZDoom/gzdoom#2337]
### GZDoom version

4.11.3

### Which game are you running with GZDoom?

Hexen

### What Operating System are you using?

Windows 10

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

From my testing, it appears that the GlassBreak() special will never jump to the other spawn states, it always spawns in at the first frame. I'm trying to set this up in Hexen with the 5 different shard frames SGSA (A,B,C,D,E). But it only ever spawns a blue shard, frame A. Here's the zscript for the GlassJunk class I'm using:

`States
	{
Spawn:
	SGSA A 128;
	Goto Death;
	SGSA B 128;
	Goto Death+1;
	SGSA C 128;
	Goto Death+2;
	SGSA D 128;
	Goto Death+3;
	SGSA E 128;
	Goto Death+4;
Death:
	SGSB A 1 A_FadeOut(0.03);
	Wait;
	SGSB B 1 A_FadeOut(0.03);
	Wait;
	SGSB C 1 A_FadeOut(0.03);
	Wait;
	SGSB D 1 A_FadeOut(0.03);
	Wait;
	SGSB E 1 A_FadeOut(0.03);
	Wait;
	}`

And the health is set to 5. Now, I might be wrong about this as I am not experienced with cpp but from what I can tell, the GlassBreak() function might have a bug, specifically this line at 3348:

`glass->SetState(glass->SpawnState + (pr_glass() % glass->health));`

It seems pr_glass() returns FRandom, which I assume to mean a random float between 0 and 1. But 0.x mod health will always return the 0.x since that's what happens when you mod a fraction, it just returns the fraction. So my assumption here is that this line of code will always return 0 (possibly rounded down or casted down to an int?), thus spawning in on the first frame which isn't the intended use of this code.

A possible solution could be something like this:

`glass->SetState(glass->SpawnState + round(pr_glass() * glass->health) );`

Which would return a random integer between 0 and health, which is the intended use of this function. 

But maybe I'm missing something. 😅

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Set line special to GlassBreak
2. Activate line special in gzdoom running hexen
3. Observe only the first frame shards will spawn

**Comment by coelckers** [comment]
pr_glass() returns an integer between 0 and 255.

**Comment by deathzero021** [comment]
I see, so the math always ends up returning 0, thus the first frame in the spawn state is always used. 

`(0~255) mod 5` will always return 0.

**Comment by RicardoLuis0** [comment]
> `(0~255) mod 5` will always return 0
no? mod is remainder, so with `mod 5`, 0 will be 0, 1 will be 1 ... 4 will be 4, 5 will be 0, 6 will be 1, and so on

**Comment by Boondorl** [comment]
`FRandom` internally is not the same as ZScript's `FRandom` function. It's a part of the questionable naming convention of structs being prefixed with `F`. By default the `()` operator returns an int from 0 to 255 in Doom fashion. Do you have the special's `arg1` set by chance? That appears to disable randomizing the spawn state.

**Comment by deathzero021** [comment]
> `FRandom` internally is not the same as ZScript's `FRandom` function. It's a part of the questionable naming convention of structs being prefixed with `F`. By default the `()` operator returns an int from 0 to 255 in Doom fashion. Do you have the special's `arg1` set by chance? That appears to disable randomizing the spawn state.

I have arg1 set to true, as the wiki says this: `If dontspawnjunk is 0, seven actors of type are spawned at the line's center (at the side the line is facing to). type is the [spawn number](https://zdoom.org/wiki/Spawn_number) associated with the actor class to spawn. If it is 0, the actor class to spawn defaults to [GlassJunk].`

And I'm using an actor inheriting and replacing GlassJunk. The correct actor is spawning, it just won't randomize the spawn frame. But I'll try experimenting with the args to see if something else is going on with them. Maybe setting the type manually will have better luck.



> > `(0~255) mod 5` will always return 0
> 
> no? mod is remainder, so with `mod 5`, 0 will be 0, 1 will be 1, ... , 4 will be 4, 5 will be 0, 6 will be 1, and so on

Haha you are correct 😅  I don't know what my early morning brain was thinking, I did some quick checks on my calculator and must've only tested increments of 5. So than I have no idea why it isn't working then, since the math does check out. 🤔

**Comment by Boondorl** [comment]
Sorry, by arg1 I meant what the wiki probably refers to as arg2 (sometimes their 0-based going from 0-4 instead of 1-5)

**Comment by deathzero021** [comment]
> Sorry, by arg1 I meant what the wiki probably refers to as arg2 (sometimes they're 0-based going from 0-4 instead of 1-5)

In Ultimate Doom Builder, it only has 1 argument "spawn glass shards" Yes or No. Setting it to No results in nothing being spawned. Setting it to Yes spawns GlassJunk (or any actor replacing it). I'm trying to test if I can spawn a different actor by passing in a spawn number for arg2, but I'm getting an error "unknown property" when using SpawnID 😅

**Comment by prof-hastig** [comment]
> Sorry, by arg1 I meant what the wiki probably refers to as arg2 (sometimes they're 0-based going from 0-4 instead of 1-5). Any value > 0 in that one will disable the randomization.

That is not correct, actually.
The second argument allows specifying a custom actor type via its spawn ID. If you set it to 0 or something non-existent, nothing will spawn, of course.
But here's the gotcha: If you do that, your actor needs to take care of selecting its spawn state itself! The automatic randomization only happens for the original GlassJunk or any replacement being used on it.

I am fairly certain that this all is just the result of a fundamental misunderstanding how this special works. What is needed here is not code fragments but a runnable example that demonstrates the reported problem. Without it all we can do is guess.

**Comment by coelckers** [comment]
I have to concur that without a runnable example this is unlikely to be resolved because I have no good idea what setup has been used

**Comment by deathzero021** [comment]
Alright... turns out I'm an idiot haha. I apologize for confusing everyone with my confusion. Firstly, I got the number of frames wrong on the glass shard sprites in Hexen.wad.  Not sure how I got that confused but I did. Secondly, I didn't realize that tics is being randomized by GlassBreak() before it enters the spawn state, which required an override to bypass that, it was causing some very unexpected behavior.

Finally, I got it working fairly decently. I made a [small example](https://www.dropbox.com/scl/fi/oeipzxw4hxze3tkdz5nfm/GlassTest.pk3?rlkey=9397iom6anmz5etdvsz7u6l4l&dl=0) showing it working properly with hexen's sprites. So the function definitely works properly as intended, it's just a little unintuitive to setup. Consider this "bug" solved 😅
